activation/nipost_state.go

```diff
@@ -20,43 +19,71 @@ const (
 	challengeFilename = "nipost_challenge.bin"
 	builderFilename   = "nipost_builder_state.bin"
 	postFilename      = "post.bin"
-	crc64Size         = 8
 )
 
 func write(path string, data []byte) error {
-	buf := bytes.NewBuffer(nil)
+	tmp, err := os.CreateTemp("", filepath.Base(path))
+	if err != nil {
+		return fmt.Errorf("create temporary file %s: %w", tmp.Name(), err)
+	}
+
 	checksum := crc64.New(crc64.MakeTable(crc64.ISO))
-	w := io.MultiWriter(buf, checksum)
-	if _, err := w.Write(data); err != nil {
-		return fmt.Errorf("write data %v: %w", path, err)
+	w := io.MultiWriter(tmp, checksum)
+	if _, err = w.Write(data); err != nil {
+		_ = tmp.Close()
+		return fmt.Errorf("write data %v: %w", tmp.Name(), err)
 	}
 
-	crc := make([]byte, crc64Size)
+	crc := make([]byte, crc64.Size)
 	binary.BigEndian.PutUint64(crc, checksum.Sum64())
-	if _, err := buf.Write(crc); err != nil {
-		return fmt.Errorf("write checksum %v: %w", path, err)
+	if _, err = tmp.Write(crc); err != nil {
+		_ = tmp.Close()
+		return fmt.Errorf("write checksum %s: %w", tmp.Name(), err)
 	}
-	if err := atomic.WriteFile(path, buf); err != nil {
-		return fmt.Errorf("save file %v: %w", path, err)
+
+	if err = tmp.Close(); err != nil {
+		return fmt.Errorf("failed to close tmp file %s: %w", tmp.Name(), err)
 	}
+
+	if err = atomic.ReplaceFile(tmp.Name(), path); err != nil {
+		return fmt.Errorf("save file from %s, %s: %w", tmp.Name(), path, err)
+	}
+
 	return nil
 }
 
 func read(path string) ([]byte, error) {
-	data, err := os.ReadFile(path)
+	file, err := os.Open(path)
 	if err != nil {
-		return nil, fmt.Errorf("read file %v: %w", path, err)
+		return nil, fmt.Errorf("open file %s: %w", path, err)
 	}
+
+	defer file.Close()
+
+	fInfo, err := file.Stat()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get file info %s: %w", path, err)
+	}
+
+	data := make([]byte, fInfo.Size()-crc64.Size)
 	checksum := crc64.New(crc64.MakeTable(crc64.ISO))
-	end := len(data) - crc64Size
-	if _, err = checksum.Write(data[:end]); err != nil {
-		return nil, fmt.Errorf("read checksum %v: %w", path, err)
+	if _, err = io.TeeReader(file, checksum).Read(data); err != nil {
+		return nil, fmt.Errorf("read file %s: %w", path, err)
+	}
+
+	saved := make([]byte, crc64.Size)
+	if _, err = file.Read(saved); err != nil {
```

Comment:
won't it fail with io unexpected eof here for existing files without crc? can we please add such test?

Reply:
The code stats the file, and first reads all data except the last 8 (`crc64.size`) bytes from the file into data. Then it reads the remaining 8 bytes into `saved` and compares if saved matches the calculated CRC.

There can only be an `io.EOF` here if the file changes size while the node is reading it 😕

Reply:
Yes, if the CRC is missing the last 8 bits will be read as CRC and then the checksum verification will fail. I'll add a quick test for that case.

If you want a test where the file is changed between the reads I can try to make one but that might be a bit trickier. Because we pass the path as a string. But if you want I can mock the filesystem for `read` and test it.